ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceCall.java

```diff
@@ -0,0 +1,251 @@
+
+public class TraceCall implements JsonRpcMethod {
+  private static final String TRACE_NODE_PROPERTY_NAME = "trace";
+  private static final String OUTPUT_NODE_PROPERTY_NAME = "output";
+  private static final String STATE_DIFF_NODE_PROPERTY_NAME = "stateDiff";
+  private static final String VM_TRACE_NODE_PROPERTY_NAME = "vmTrace";
+
+  private final Supplier<BlockchainQueries> blockchainQueries;
+  private final ProtocolSchedule protocolSchedule;
+  private final TransactionSimulator transactionSimulator;
+
+  public TraceCall(
+      final BlockchainQueries blockchainQueries,
+      final ProtocolSchedule protocolSchedule,
+      final TransactionSimulator transactionSimulator) {
+    this.blockchainQueries = Suppliers.ofInstance(blockchainQueries);
+    this.protocolSchedule = protocolSchedule;
+    this.transactionSimulator = transactionSimulator;
+  }
+
+  @Override
+  public String getName() {
+    return transactionSimulator != null ? RpcMethod.TRACE_CALL.getMethodName() : null;
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    final TraceTypeParameter traceTypeParameter =
+        requestContext.getRequiredParameter(1, TraceTypeParameter.class);
+    final Optional<BlockParameter> maybeBlockParameter =
+        requestContext.getOptionalParameter(2, BlockParameter.class);
+
+    final Optional<BlockHeader> maybeBlockHeader = resolveBlockHeader(maybeBlockParameter);
+
+    if (maybeBlockHeader.isEmpty()) {
+      throw new IllegalStateException("Invalid block parameter.");
+    }
+
+    final Set<TraceTypeParameter.TraceType> traceTypes = traceTypeParameter.getTraceTypes();
+    final DebugOperationTracer tracer = new DebugOperationTracer(buildTraceOptions(traceTypes));
+    final Optional<TransactionSimulatorResult> maybeSimulatorResult =
+        transactionSimulator.process(
+            JsonCallParameterUtil.validateAndGetCallParams(requestContext),
+            buildTransactionValidationParams(),
+            tracer,
+            maybeBlockHeader.get());
+
+    if (maybeSimulatorResult.isEmpty()) {
+      throw new IllegalStateException("Invalid transaction simulator result.");
+    }
+
+    final ObjectMapper mapper = new ObjectMapper();
+    final ObjectNode resultNode = mapper.createObjectNode();
+
+    resultNode.put(OUTPUT_NODE_PROPERTY_NAME, maybeSimulatorResult.get().getOutput().toString());
+
+    final TransactionTrace transactionTrace =
+        new TransactionTrace(
+            maybeSimulatorResult.get().getTransaction(),
+            maybeSimulatorResult.get().getResult(),
+            tracer.getTraceFrames());
+
+    final Block block = blockchainQueries.get().getBlockchain().getChainHeadBlock();
+    final AtomicInteger traceCounter = new AtomicInteger(0);
+
+    if (traceTypes.contains(TraceTypeParameter.TraceType.STATE_DIFF)) {
+      final StateDiffGenerator stateDiffGenerator = new StateDiffGenerator();
+
+      generateTracesFromTransactionTrace(
```

Comment:
I see the same code in the TraceReplayBlockTransactions class. Is there no way to avoid duplication? Maybe not I'm just wondering

Reply:
I've refactored the code based on [Gary's advice](https://github.com/hyperledger/besu/pull/3166#discussion_r797891461). Now there is no more code duplication